Python/pythonrun.c

```diff
@@ -1366,7 +1366,7 @@ Py_CompileStringObject(const char *str, PyObject *filename, int start,
         return NULL;
 
     mod = _PyParser_ASTFromString(str, filename, start, flags, arena);
-    if (mod == NULL) {
+    if (mod == NULL || !_PyAST_Validate(mod)) {
```

Comment:
Wait, why is this change here? THis is going to slow down quite a lot `Py_CompileStringObject` code path

Reply:
Let's remove it for now, then. I understood it as being motivated by the points in [this comment](https://github.com/python/cpython/pull/25585#issuecomment-826926615).

Reply:
Regading this comment there is an important fact to highlight:

> compiling from strings didn't attempt to validate them because it never attempted to validate anything

The actual reason is because by design the parser code *must* return a valid AST (there is a debug-mode validation step to ensure that this contract is true).

Reply:
Hm, looks like the new code might have been relying on that pass to give us a SyntaxError in `test_patma_240`. I'll try to figure out what's going on.

Reply:
The parser can't verify that a complex literal is actually a complex literal because the tokeniser just emits `NUMBER` for both real and imaginary numbers. As a result, `0+0` makes it through the parsing rules for pattern matching.

The compiler can't reject the syntax because it never sees it - the value has turned into the legal constant `0` by the time the pattern gets to it.

The original pattern matching implementation worked around that limitation by putting a validation step in the AST constant folding pass: instead of using the standard constant folding for expressions, it reimplemented the subset required to fold the literals that pattern matching supported and emitted a syntax error for binary operations that weren't complex numbers.

This PR switches over to using the standard expression folding functions in the optimisation pass, so forcing validation was needed to keep "0+0" illegal in value patterns and mapping pattern keys.

I'll try replace it with a more targeted fix that adds a helper function to pegen that ensures the AST coming out of the parser is valid rather than relying on the AST validator or the AST constant folding to reject it.

Reply:
> The compiler can't reject the syntax because it never sees it - the value has turned into the legal constant 0 by the time the pattern gets to it.

We probably shouldn't fold the AST for patterns in that case so the compiler can handle it. Or I am missing something?

Reply:
I just pushed a change that restores the invariant of the parser always producing a valid AST without having to weaken the checks on the AST nodes for value patterns and mapping patterns keys.

The trick is a new `_PyAST_EnsureImaginary` helper function that accepts and returns `expr_ty`.

It looks like I didn't get the reversion/merge with Brandt's changes quite right though, so I'll post an example after cleaning that up.

Reply:
I had reverted the validator and compiler changes, but the AST optimiser changes needed to be reverted as well.
This is how compilation from a string behaves with the parser enhanced to ensure that complex literals actually are complex numbers:
```
>>> match x:
...     case 0+0:
SyntaxError: Imaginary number required in complex literal
```

So with that everything is handling the responsibilities it should be handling:
* parser rejects "0+0" when compiling from a string rather than producing an invalid AST
* validator rejects "0+0" when compiling from a prebuilt AST tree
* optimiser and compiler have enough checks to protect themselves from segfaults, but are otherwise pretty tolerant

Reply:
After the first iteration, I realised the imaginary number checking would work better as pegen helper rather than as an AST helper:
```
>>> match x:
...     case 0+0:
  File "<stdin>", line 2
    case 0+0:
           ^
SyntaxError: Imaginary number required in complex literal
```